scheduler/workflow.go

```diff
@@ -335,26 +336,38 @@ func (s *schedulerWorkflow) StateString() string {
 }
 
 func (s *schedulerWorkflow) workJobs(prs []*processNode, pus []*publishNode, t *task, pj job) {
+	// create a waitgroup track pending goroutines submitting jobs
+	wg := &sync.WaitGroup{}
 	for _, pr := range prs {
-		j := newProcessJob(pj, pr.Name(), pr.Version(), pr.InboundContentType, pr.config.Table(), t.metricsManager, t.id)
-		errors := t.manager.Work(j).Promise().Await()
-		if len(errors) != 0 {
-			t.failedRuns++
-			t.lastFailureTime = t.lastFireTime
-			t.lastFailureMessage = errors[len(errors)-1].Error()
-			return
-		}
+		go func() {
+			wg.Add(1)
+			defer wg.Done()
+			j := newProcessJob(pj, pr.Name(), pr.Version(), pr.InboundContentType, pr.config.Table(), t.metricsManager, t.id)
+			errors := t.manager.Work(j).Promise().Await()
+			if len(errors) != 0 {
+				t.failedRuns++
```

Comment:
The mutex on `t (scheduler.Task)` is used for state change safety:
https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L207
https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L218
https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L228
https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L240
https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L335
https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L296
https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L309
https://github.com/intelsdi-x/snap/blob/master/scheduler/task.go#L325

In fact, the above makes me think the internals of task should have a private function for changing state and unexport the sync.Mutex.

Since we are are just sending jobs into the task queue effectively, this does not conflict with state change since it never changes the state directly. Effectively this use of `t` is only for calling Work(j) and any possible `t.state` change is not a thread safety problem.

However, one could reason that job queues should be cleared if `t.state` were to switch to an ended or disabled state. This is right now not being done safely. I think we may want a job submission mutex to synchronize the clearing of the queue of jobs for a task with new submissions that may (or may not) be allowed.